packages/mui-system/src/Container/createContainer.tsx

```diff
@@ -0,0 +1,185 @@
+import { unstable_capitalize as capitalize } from '@mui/utils';
+import { OverridableComponent } from '@mui/types';
+import {
+  unstable_composeClasses as composeClasses,
+  generateUtilityClass,
+} from '@mui/private-classnames';
+import { ContainerProps, ContainerTypeMap } from './ContainerProps';
+import useThemePropsSystem from '../useThemeProps';
+import systemStyled from '../styled';
+import createTheme, { Theme as DefaultTheme, Breakpoint } from '../createTheme';
+
+interface StyleFnProps<Theme> extends ContainerProps {
+  theme: Theme;
+  ownerState: ContainerProps;
+}
+
+const defaultTheme = createTheme();
+
+const defaultCreateStyledComponent = systemStyled('div', {
+  name: 'MuiContainer',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    const { ownerState } = props;
+
+    return [
+      styles.root,
+      styles[`maxWidth${capitalize(String(ownerState.maxWidth))}`],
+      ownerState.fixed && styles.fixed,
+      ownerState.disableGutters && styles.disableGutters,
+    ];
+  },
+});
+
+const useThemePropsDefault = (inProps: ContainerProps) =>
+  useThemePropsSystem({ props: inProps, name: 'MuiContainer', defaultTheme });
+
+const useUtilityClasses = (ownerState: ContainerProps, componentName: string) => {
+  const getContainerUtilityClass = (slot: string) => {
+    return generateUtilityClass(componentName, slot);
+  };
+  const { classes, fixed, disableGutters, maxWidth } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      maxWidth && `maxWidth${capitalize(String(maxWidth))}`,
+      fixed && 'fixed',
+      disableGutters && 'disableGutters',
+    ],
+  };
+
+  return composeClasses(slots, getContainerUtilityClass, classes);
+};
+
+type RequiredThemeStructure = Pick<DefaultTheme, 'breakpoints' | 'spacing'>;
+
+export default function createContainer<Theme extends RequiredThemeStructure = DefaultTheme>(
+  options: {
+    createStyledComponent?: (
+      ...styles: Array<Interpolation<StyleFnProps<Theme>>>
+    ) => StyledComponent<ContainerProps>;
+    useThemeProps?: (inProps: ContainerProps) => ContainerProps & { component?: React.ElementType };
+    componentName?: string;
+  } = {},
+) {
+  const {
+    // This will allow adding custom styled fn (for example for custom sx style function)
+    createStyledComponent = defaultCreateStyledComponent,
+    useThemeProps = useThemePropsDefault,
+    componentName = 'MuiContainer',
+  } = options;
+
+  const ContainerRoot = createStyledComponent(
+    ({ theme, ownerState }: StyleFnProps<Theme>) =>
+      ({
+        width: '100%',
+        marginLeft: 'auto',
+        boxSizing: 'border-box',
+        marginRight: 'auto',
+        display: 'block', // Fix IE11 layout when used with main.
+        ...(!ownerState.disableGutters && {
+          paddingLeft: theme.spacing(2),
+          paddingRight: theme.spacing(2),
+          // @ts-ignore module augmentation fails if custom breakpoints are used
```

Comment:
I had to add bunch of @ts-ignore when using breakpoints, as we have a module augmentation test that fails as these breakpoints do not exits 😅 Not the best, but I don't have a better option for now...